keps/sig-api-machinery/4192-svm-in-tree/README.md

```diff
@@ -0,0 +1,483 @@
+- Any modification regarding `StorageVersion` API for HA API servers
+- Adding logic that relies on the hashed storage versions exposed via the discovery API
+
+### UNCLEAR Goals and/or Non-Goals
+
+- Automatic storage version migration for CRDs
+- Make it easy for Kubernetes developers to drop old API schemas by guaranteeing that storage migration is run automatically on SV hash changes (should this also be on a timer or API server identity?)
+- Automated Storage Version Migration via the hash exposed by the `StorageVersion` API
+
+## Proposal
+
+- Move the existing SVM controller logic in-tree into KCM
+- Move the existing SVM REST APIs in-tree (possibly under a new API group to avoid conflicts with the old API being run concurrently)
+- **[UNCLEAR]** Create a new controller that watches the storage version API and automatically triggers migrations based on changes (to the SV API)
+
+### User Stories (Optional)
+
+#### Story 1 **[UNCLEAR]**
+As an end user of Kubernetes, we get automatic storage version migration whenever the storage version changes due to an api server upgrade/downgrade.
+
+#### Story 2
+As an end user using encryption at rest, whenever the key change is detected we can run the storage migration to use the new key for encryption. 
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+## Design Details
+
+### APIs to move
+#### We will move following [APIs](https://github.com/kubernetes-sigs/kube-storage-version-migrator/blob/60dee538334c2366994c2323c0db5db8ab4d2838/pkg/apis/migration/v1alpha1/types.go) in-tree:
+- `v1alpha1` of `storageversionmigrations.migration.k8s.io`  
+
+    ```go
+    // StorageVersionMigration represents a migration of stored data to the latest
+    // storage version.
+    type StorageVersionMigration struct {
+        metav1.TypeMeta `json:",inline"`
+        // +optional
+        metav1.ObjectMeta `json:"metadata,omitempty"`
+        // Specification of the migration.
+        // +optional
+        Spec StorageVersionMigrationSpec `json:"spec,omitempty"`
+        // Status of the migration.
+        // +optional
+        Status StorageVersionMigrationStatus `json:"status,omitempty"`
+    }
+
+    // Spec of the storage version migration.
+    type StorageVersionMigrationSpec struct {
+        // The resource that is being migrated. The migrator sends requests to
+        // the endpoint serving the resource.
+        // Immutable.
+        Resource GroupVersionResource `json:"resource"`
+        // The token used in the list options to get the next chunk of objects
+        // to migrate. When the .status.conditions indicates the migration is
+        // "Running", users can use this token to check the progress of the
+        // migration.
+        // +optional
+        ContinueToken string `json:"continueToken,omitempty"`
+        // The storage version hash to avoid races.
+        StorageVersionHash string `json:"storageVersionHash,omitempty"`
```

Comment:
Sorry if this is obvious, but what's this for?

Reply:
I added it to resolve - https://github.com/kubernetes-sigs/kube-storage-version-migrator/blob/60dee538334c2366994c2323c0db5db8ab4d2838/pkg/apis/migration/v1alpha1/types.go#L64-L65

If I understand this correctly, we can use the hash to validate whether the resource still needs migration.

However, upon further examination, I realize that in the out-of-tree SVM migration, the Migration Resource is created automatically, making it easier to populate this field. With the current implementation, we are asking the user to create the Migration Resource. This may not provide the best user experience, as users would need to know the appropriate hash. WDYT?

cc @enj